File: tide_notifier.js
// tide_notifier.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const { program } = require('commander');

const CONFIG_FILE = 'tide_notifier_config.json';
const DEFAULT_INTERVAL = 60;
const DEFAULT_THRESHOLD = 30;

class TideModel {
    static level(hours, phaseOffset = 0) {
        const M2_PERIOD = 12.42;
        const S2_PERIOD = 12.0;
        const radM2 = (hours / M2_PERIOD + phaseOffset / M2_PERIOD) * 2 * Math.PI;
        const radS2 = (hours / S2_PERIOD + phaseOffset / S2_PERIOD) * 2 * Math.PI;
        return 0.0 + 1.2 * Math.sin(radM2) + 0.4 * Math.sin(radS2);
    }

    static nextEvent(hours, phaseOffset = 0) {
        const step = 0.05;
        const times = [];
        const levels = [];
        for (let i = 0; i < Math.floor(12 * 60 / (step * 60)) + 1; i++) {
            const t = hours + i * step;
            times.push(t);
            levels.push(this.level(t, phaseOffset));
        }
        let highTimes = [], lowTimes = [];
        for (let i = 1; i < times.length - 1; i++) {
            if (levels[i] > levels[i-1] && levels[i] > levels[i+1]) {
                highTimes.push({t: times[i], lvl: levels[i]});
            } else if (levels[i] < levels[i-1] && levels[i] < levels[i+1]) {
                lowTimes.push({t: times[i], lvl: levels[i]});
            }
        }
        let nextHigh = null, nextLow = null;
        for (const h of highTimes) {
            if (h.t >= hours) {
                nextHigh = h;
                break;
            }
        }
        if (!nextHigh && highTimes.length > 0) {
            nextHigh = {t: highTimes[0].t + 24, lvl: highTimes[0].lvl};
        }
        for (const l of lowTimes) {
            if (l.t >= hours) {
                nextLow = l;
                break;
            }
        }
        if (!nextLow && lowTimes.length > 0) {
            nextLow = {t: lowTimes[0].t + 24, lvl: lowTimes[0].lvl};
        }
        return { nextHigh, nextLow };
    }
}

class Notifier {
    constructor() {
        this.locations = [];
        this.load();
    }

    load() {
        if (fs.existsSync(CONFIG_FILE)) {
            try {
                this.locations = JSON.parse(fs.readFileSync(CONFIG_FILE));
            } catch (e) {}
        }
    }

    save() {
        fs.writeFileSync(CONFIG_FILE, JSON.stringify(this.locations, null, 2));
    }

    addLocation(name, phaseOffset, threshold = DEFAULT_THRESHOLD) {
        this.locations = this.locations.filter(l => l.name !== name);
        this.locations.push({ name, phase_offset: phaseOffset, threshold });
        this.save();
        console.log(`✅ Location '${name}' added.`);
    }

    removeLocation(name) {
        this.locations = this.locations.filter(l => l.name !== name);
        this.save();
        console.log(`✅ Location '${name}' removed.`);
    }

    listLocations() {
        if (!this.locations.length) {
            console.log('No locations.');
            return;
        }
        console.log('\n📍 Monitored locations:');
        for (const loc of this.locations) {
            console.log(`  ${loc.name} (phase: ${loc.phase_offset.toFixed(2)}h, threshold: ${loc.threshold} min)`);
        }
    }

    getStatus(hours) {
        return this.locations.map(loc => {
            const lvl = TideModel.level(hours, loc.phase_offset);
            const { nextHigh, nextLow } = TideModel.nextEvent(hours, loc.phase_offset);
            let nextType = '', nextTime = null, nextLevel = null;
            if (nextHigh && nextLow) {
                if (nextHigh.t < nextLow.t) {
                    nextType = 'High';
                    nextTime = nextHigh.t;
                    nextLevel = nextHigh.lvl;
                } else {
                    nextType = 'Low';
                    nextTime = nextLow.t;
                    nextLevel = nextLow.lvl;
                }
            } else if (nextHigh) {
                nextType = 'High';
                nextTime = nextHigh.t;
                nextLevel = nextHigh.lvl;
            } else if (nextLow) {
                nextType = 'Low';
                nextTime = nextLow.t;
                nextLevel = nextLow.lvl;
            }
            return { loc, lvl, nextType, nextTime, nextLevel };
        });
    }

    monitor(interval = DEFAULT_INTERVAL) {
        console.log(`🌊 Tide Notifier (checking every ${interval}s)`);
        console.log('Press Ctrl+C to stop.\n');
        const timer = setInterval(() => {
            const now = new Date();
            const hours = now.getHours() + now.getMinutes()/60 + now.getSeconds()/3600;
            const statuses = this.getStatus(hours);
            for (const st of statuses) {
                if (st.nextType) {
                    let minutesUntil = (st.nextTime - hours) * 60;
                    if (minutesUntil < 0) minutesUntil += 24*60;
                    if (minutesUntil <= st.loc.threshold && minutesUntil > 0) {
                        console.log(`[${now.toISOString().slice(0,16).replace('T',' ')}] ${st.loc.name} – ${st.nextType} tide in ${Math.floor(minutesUntil)} min (level: ${st.nextLevel.toFixed(2)} m)`);
                    }
                }
            }
        }, interval * 1000);

        process.on('SIGINT', () => {
            clearInterval(timer);
            console.log('\nMonitoring stopped.');
            process.exit();
        });
    }

    showStatus() {
        const now = new Date();
        const hours = now.getHours() + now.getMinutes()/60 + now.getSeconds()/3600;
        const statuses = this.getStatus(hours);
        console.log(`\n🌊 Tide Status (${now.toISOString().slice(0,16).replace('T',' ')})`);
        for (const st of statuses) {
            console.log(`\n📍 ${st.loc.name}`);
            console.log(`  Phase offset: ${st.loc.phase_offset.toFixed(2)}h`);
            console.log(`  Current level: ${st.lvl.toFixed(2)} m`);
            if (st.nextType) {
                let minutesUntil = (st.nextTime - hours) * 60;
                if (minutesUntil < 0) minutesUntil += 24*60;
                const hour = Math.floor(st.nextTime);
                const min = Math.floor((st.nextTime - hour) * 60);
                console.log(`  Next event: ${st.nextType} tide at ${String(hour).padStart(2,'0')}:${String(min).padStart(2,'0')} (${st.nextLevel.toFixed(2)} m) – in ${Math.floor(minutesUntil)} min`);
            }
            console.log(`  Alert threshold: ${st.loc.threshold} min`);
        }
    }
}

program
    .command('add <name> <phase_offset>')
    .option('--threshold <minutes>', 'Alert threshold', parseInt, DEFAULT_THRESHOLD)
    .action((name, phaseOffset, options) => {
        const notifier = new Notifier();
        notifier.addLocation(name, parseFloat(phaseOffset), options.threshold);
    });

program
    .command('list')
    .action(() => {
        const notifier = new Notifier();
        notifier.listLocations();
    });

program
    .command('remove <name>')
    .action((name) => {
        const notifier = new Notifier();
        notifier.removeLocation(name);
    });

program
    .command('start')
    .option('--interval <seconds>', 'Check interval', parseInt, DEFAULT_INTERVAL)
    .action((options) => {
        const notifier = new Notifier();
        notifier.monitor(options.interval);
    });

program
    .command('status')
    .action(() => {
        const notifier = new Notifier();
        notifier.showStatus();
    });

program.parse(process.argv);

File: tide_notifier.go
// tide_notifier.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"math"
	"os"
	"os/signal"
	"syscall"
	"time"
)

const (
	DefaultInterval  = 60
	DefaultThreshold = 30
	ConfigFile       = "tide_notifier_config.json"
)

type Location struct {
	Name         string  `json:"name"`
	PhaseOffset  float64 `json:"phase_offset"`
	Threshold    int     `json:"threshold"`
}

type TideModel struct{}

func (t TideModel) Level(hours, phaseOffset float64) float64 {
	m2Period := 12.42
	s2Period := 12.0
	radM2 := (hours/m2Period + phaseOffset/m2Period) * 2 * math.Pi
	radS2 := (hours/s2Period + phaseOffset/s2Period) * 2 * math.Pi
	return 0.0 + 1.2*math.Sin(radM2) + 0.4*math.Sin(radS2)
}

func (t TideModel) NextEvent(hours, phaseOffset float64) (nextHigh, nextLow struct{ t, lvl float64 }) {
	step := 0.05
	var times []float64
	var levels []float64
	for i := 0; i < int(12*60/(step*60))+1; i++ {
		tt := hours + float64(i)*step
		times = append(times, tt)
		levels = append(levels, t.Level(tt, phaseOffset))
	}
	var highTimes, lowTimes []struct{ t, lvl float64 }
	for i := 1; i < len(times)-1; i++ {
		if levels[i] > levels[i-1] && levels[i] > levels[i+1] {
			highTimes = append(highTimes, struct{ t, lvl float64 }{times[i], levels[i]})
		} else if levels[i] < levels[i-1] && levels[i] < levels[i+1] {
			lowTimes = append(lowTimes, struct{ t, lvl float64 }{times[i], levels[i]})
		}
	}
	// Find next from now
	for _, h := range highTimes {
		if h.t >= hours {
			nextHigh = h
			break
		}
	}
	if nextHigh.t == 0 && len(highTimes) > 0 {
		nextHigh = highTimes[0]
		nextHigh.t += 24
	}
	for _, l := range lowTimes {
		if l.t >= hours {
			nextLow = l
			break
		}
	}
	if nextLow.t == 0 && len(lowTimes) > 0 {
		nextLow = lowTimes[0]
		nextLow.t += 24
	}
	return
}

type Notifier struct {
	Locations []Location `json:"locations"`
}

func (n *Notifier) load() {
	data, err := os.ReadFile(ConfigFile)
	if err != nil {
		return
	}
	json.Unmarshal(data, n)
}

func (n *Notifier) save() {
	data, _ := json.MarshalIndent(n, "", "  ")
	os.WriteFile(ConfigFile, data, 0644)
}

func (n *Notifier) AddLocation(name string, phaseOffset float64, threshold int) {
	if threshold == 0 {
		threshold = DefaultThreshold
	}
	// Remove if exists
	var newLocs []Location
	for _, loc := range n.Locations {
		if loc.Name != name {
			newLocs = append(newLocs, loc)
		}
	}
	newLocs = append(newLocs, Location{Name: name, PhaseOffset: phaseOffset, Threshold: threshold})
	n.Locations = newLocs
	n.save()
	fmt.Printf("✅ Location '%s' added.\n", name)
}

func (n *Notifier) RemoveLocation(name string) {
	var newLocs []Location
	for _, loc := range n.Locations {
		if loc.Name != name {
			newLocs = append(newLocs, loc)
		}
	}
	n.Locations = newLocs
	n.save()
	fmt.Printf("✅ Location '%s' removed.\n", name)
}

func (n *Notifier) ListLocations() {
	if len(n.Locations) == 0 {
		fmt.Println("No locations.")
		return
	}
	fmt.Println("\n📍 Monitored locations:")
	for _, loc := range n.Locations {
		fmt.Printf("  %s (phase: %.2fh, threshold: %d min)\n", loc.Name, loc.PhaseOffset, loc.Threshold)
	}
}

func (n *Notifier) getStatus(hours float64) []struct {
	loc       Location
	lvl       float64
	nextType  string
	nextTime  float64
	nextLevel float64
} {
	var statuses []struct {
		loc       Location
		lvl       float64
		nextType  string
		nextTime  float64
		nextLevel float64
	}
	tm := TideModel{}
	for _, loc := range n.Locations {
		lvl := tm.Level(hours, loc.PhaseOffset)
		high, low := tm.NextEvent(hours, loc.PhaseOffset)
		var nextType string
		var nextTime, nextLevel float64
		if high.t > 0 && low.t > 0 {
			if high.t < low.t {
				nextType = "High"
				nextTime = high.t
				nextLevel = high.lvl
			} else {
				nextType = "Low"
				nextTime = low.t
				nextLevel = low.lvl
			}
		} else if high.t > 0 {
			nextType = "High"
			nextTime = high.t
			nextLevel = high.lvl
		} else if low.t > 0 {
			nextType = "Low"
			nextTime = low.t
			nextLevel = low.lvl
		}
		statuses = append(statuses, struct {
			loc       Location
			lvl       float64
			nextType  string
			nextTime  float64
			nextLevel float64
		}{loc, lvl, nextType, nextTime, nextLevel})
	}
	return statuses
}

func (n *Notifier) Monitor(interval int) {
	fmt.Printf("🌊 Tide Notifier (checking every %ds)\n", interval)
	fmt.Println("Press Ctrl+C to stop.\n")
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)

	ticker := time.NewTicker(time.Duration(interval) * time.Second)
	go func() {
		for range ticker.C {
			now := time.Now()
			hours := float64(now.Hour()) + float64(now.Minute())/60.0 + float64(now.Second())/3600.0
			statuses := n.getStatus(hours)
			for _, st := range statuses {
				if st.nextType != "" {
					minutesUntil := (st.nextTime - hours) * 60
					if minutesUntil < 0 {
						minutesUntil += 24 * 60
					}
					if minutesUntil <= float64(st.loc.Threshold) && minutesUntil > 0 {
						fmt.Printf("[%s] %s – %s tide in %.0f min (level: %.2f m)\n",
							now.Format("2006-01-02 15:04"), st.loc.Name, st.nextType, minutesUntil, st.nextLevel)
					}
				}
			}
		}
	}()
	<-sigChan
	fmt.Println("\nMonitoring stopped.")
	ticker.Stop()
}

func (n *Notifier) ShowStatus() {
	now := time.Now()
	hours := float64(now.Hour()) + float64(now.Minute())/60.0 + float64(now.Second())/3600.0
	statuses := n.getStatus(hours)
	fmt.Printf("\n🌊 Tide Status (%s)\n", now.Format("2006-01-02 15:04"))
	for _, st := range statuses {
		fmt.Printf("\n📍 %s\n", st.loc.Name)
		fmt.Printf("  Phase offset: %.2fh\n", st.loc.PhaseOffset)
		fmt.Printf("  Current level: %.2f m\n", st.lvl)
		if st.nextType != "" {
			minutesUntil := (st.nextTime - hours) * 60
			if minutesUntil < 0 {
				minutesUntil += 24 * 60
			}
			hour := int(st.nextTime)
			min := int((st.nextTime - float64(hour)) * 60)
			fmt.Printf("  Next event: %s tide at %02d:%02d (%.2f m) – in %.0f min\n",
				st.nextType, hour, min, st.nextLevel, minutesUntil)
		}
		fmt.Printf("  Alert threshold: %d min\n", st.loc.Threshold)
	}
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: tide_notifier <command> [options]")
		return
	}
	notifier := &Notifier{}
	notifier.load()

	cmd := os.Args[1]
	switch cmd {
	case "add":
		addCmd := flag.NewFlagSet("add", flag.ExitOnError)
		name := addCmd.String("name", "", "")
		phase := addCmd.Float64("phase", 0, "")
		threshold := addCmd.Int("threshold", 0, "")
		addCmd.Parse(os.Args[2:])
		// parse positional args
		args := addCmd.Args()
		if len(args) < 2 {
			fmt.Println("Usage: add <name> <phase_offset> [--threshold <minutes>]")
			return
		}
		nameStr := args[0]
		phaseStr := args[1]
		phaseVal, _ := strconv.ParseFloat(phaseStr, 64)
		thr := *threshold
		if thr == 0 {
			thr = DefaultThreshold
		}
		notifier.AddLocation(nameStr, phaseVal, thr)
	case "list":
		notifier.ListLocations()
	case "remove":
		if len(os.Args) < 3 {
			fmt.Println("remove <name>")
			return
		}
		notifier.RemoveLocation(os.Args[2])
	case "start":
		startCmd := flag.NewFlagSet("start", flag.ExitOnError)
		interval := startCmd.Int("interval", DefaultInterval, "Check interval in seconds")
		startCmd.Parse(os.Args[2:])
		notifier.Monitor(*interval)
	case "status":
		notifier.ShowStatus()
	default:
		fmt.Println("Unknown command. Available: add, list, remove, start, status")
	}
}
